ceph.spec.in: bump gcc-toolset to 13 and use it on rhel>=8

tchaikov opened the pull request:
since gts 13 is out, and the newer the toolchain the merrier. and because clang prefers using a newer gts when its gcc is around. so let's bump up the gcc-toolset from 11 to 13.. both RHEL8 and RHEL9 have gcc-toolset 13, so let's enable it on
rhel>=8, more future-proof this way.

<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

athanatos reviewed (approved):
gcc-13 also seems to have sufficiently good coroutine support to be usable.  This seems simpler than switching to clang-16/17.

@rzarzynski @cbodley Does this look ok to  you?

tchaikov commented:
changelog

- do not install `-annobin` subpackage anymore, as it does not exist in gts-13.

tchaikov commented:
@athanatos Hi Sam, FWIW, Scylladb didn't switched to GCC because of https://gcc.gnu.org/bugzilla/show_bug.cgi?id=98056, but it was fixed one year ago.

tchaikov commented:
```
-- Performing Test Cxx_Compiler_BZ107852_Free
CMake Error at /home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/19.0.0-1808-g79ff8417/rpm/el8/BUILD/ceph-19.0.0-1808-g79ff8417/x86_64-redhat-linux-gnu/CMakeFiles/CMakeScratch/TryCompile-lZ77IE/CMakeLists.txt:26 (target_link_libraries):
  Target "cmTC_2649c" links to:

    fmt::fmt

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.

```

see https://jenkins.ceph.com/job/ceph-dev-new-build/ARCH=x86_64,AVAILABLE_ARCH=x86_64,AVAILABLE_DIST=centos8,DIST=centos8,MACHINE_SIZE=gigantic/77261//consoleFull.

strange enough: `fmt::fmt` was not available when Seastar was checking for https://gcc.gnu.org/bugzilla/show_bug.cgi?id=107852

tchaikov commented:
see also https://github.com/ceph/seastar/commit/d382f247629a7349afc0cea495a1cd0502dc0b22

tchaikov commented:
i am proposing a solution at https://github.com/ceph/seastar/commit/d382f247629a7349afc0cea495a1cd0502dc0b22#r139292281

athanatos commented:
Pushed to ci as wip-sjust-testing-2024-03-04

tchaikov commented:
> Pushed to ci as wip-sjust-testing-2024-03-04

@athanatos i was testing it in https://shaman.ceph.com/builds/ceph/wip-pr-55886-kefu/8352c8cc4786283da2db0240841e2e5500699e02/crimson/384525/ as well. but i hit a FTBFS addressed by #55905. probably you could cherrypick #55905 in your branch?

tchaikov commented:
pushed to https://shaman.ceph.com/builds/ceph/wip-pr-55886-kefu/df11d2889079b3d1d933fe65c0c453b8911011b8/

tchaikov commented:
now blocked by #55910

cbodley commented:
> @rzarzynski @cbodley Does this look ok to you?

👍👍 would this allow us to drop the 'crimson' build flavor in shaman?

tchaikov commented:
this is a great question indeed. i think the answer is "YES! but ...". 

- YES: see the builds at https://shaman.ceph.com/builds/ceph/wip-pr-55886-kefu/df11d2889079b3d1d933fe65c0c453b8911011b8/ . 
- but: the reason of "crimson" flavor is two-folded: 
  * compile time dependencies: the last bits are addressed by this change
  * run time problems: ceph-osd versus crimson-osd. we don't have a solution yet to package crimson-osd in a dedicated package. what we have is to package crimson-osd with the name of `ceph-osd` (@rzarzynski do you still recall the discussion we had on this couple years ago? =))

cbodley commented:
> so i think we cannot retire the crimson flavor yet.

thanks @tchaikov. i've just been feeling the weight of our build matrix lately. the centos8 and ubuntu focal builds should get retired soonish. it would be great if we could find a way to prioritize the crimson part after squid

athanatos commented:
I think the best way to unify the packages would be to have a single ceph-osd "binary" which chooses classic-osd or crimson-osd based on a command line flag or config.

athanatos commented:
https://github.com/ceph/ceph/pull/55910 has merged.  Pushed wip-sjust-testing-2024-03-06 to ci with this PR.

athanatos commented:
The centos 8 build failed with

```
Error: 
 Problem 1: conflicting requests
  - nothing provides libprotobuf.so.30()(64bit) needed by ceph-osd-2:19.0.0-1910.g8d113723.el8.x86_64 from Ceph
 Problem 2: package ceph-volume-2:19.0.0-1910.g8d113723.el8.noarch from Ceph-noarch requires ceph-osd = 2:19.0.0-1910.g8d113723.el8, but none of the providers can be installed
  - conflicting requests
  - nothing provides libprotobuf.so.30()(64bit) needed by ceph-osd-2:19.0.0-1910.g8d113723.el8.x86_64 from Ceph
```

but I've been seeing that intermittently without this PR.  Scheduling a rados run.

athanatos commented:
http://pulpito.ceph.com/sjust-2024-03-07_23:09:46-rados-wip-sjust-testing-2024-03-06-distro-default-smithi/

athanatos commented:
I'm actually ambivalent as to whether we need to wait on the above test suite to pass -- it's possible but pretty unlikely that we'll uncover gcc-13 related compile bugs in classic and I'm not really worried about compiler bugs in crimson.  If else agrees, we can probably go ahead and merge it.

tchaikov commented:
> I think the best way to unify the packages would be to have a single ceph-osd which execs classic-osd or crimson-osd based on a command line flag or config.

wondering if we can parameterize the ceph-osd systemd service unit, so that it can launch crimson or the classic osd based on user's preference. but we need to persist a mark so user cannot switch to another osd  the the store is initialized. as the disk layouts of crimson and bluestore are different.

tchaikov commented:
> I'm actually ambivalent as to whether we need to wait on the above test suite to pass -- it's possible but pretty unlikely that we'll uncover gcc-13 related compile bugs in classic and I'm not really worried about compiler bugs in crimson. If someone else agrees, we can probably go ahead and merge it.

i tentative agree =) probably we can wait for at least a single test to pass. just to assure that the executables emitted by the new compiler and linker are sane..

tchaikov commented:
i see they are been queued. probably could single one of them with a higher priority so we can see the test result of it tomorrow.

athanatos commented:
Seems there is a problem after all.  Four failures in https://pulpito.ceph.com/sjust-2024-03-08_00:31:28-rados-wip-sjust-testing-2024-03-06-distro-default-smithi/.

3/4 look like:

```
2024-03-08T01:00:45.455 INFO:tasks.ceph:Setting crush tunables to jewel
2024-03-08T01:00:45.455 DEBUG:teuthology.orchestra.run.smithi012:> sudo ceph --cluster ceph osd crush tunables jewel
2024-03-08T01:00:45.505 INFO:tasks.ceph.mon.a.smithi012.stderr:*** Caught signal (Segmentation fault) **
2024-03-08T01:00:45.505 INFO:tasks.ceph.mon.a.smithi012.stderr: in thread 7ff093141640 thread_name:ms_dispatch
2024-03-08T01:00:45.505 INFO:tasks.ceph.mon.c.smithi012.stderr:*** Caught signal (Segmentation fault) **
2024-03-08T01:00:45.505 INFO:tasks.ceph.mon.c.smithi012.stderr: in thread 7fcc72ed6640 thread_name:ms_dispatch
2024-03-08T01:00:45.506 INFO:tasks.ceph.mon.c.smithi012.stderr: ceph version 19.0.0-1910-g8d113723 (8d113723f96506c76b2d749e6808af82c9613dd5) squid (dev)
2024-03-08T01:00:45.506 INFO:tasks.ceph.mon.c.smithi012.stderr: 1: /lib64/libc.so.6(+0x54db0) [0x7fcc7aa54db0]
2024-03-08T01:00:45.506 INFO:tasks.ceph.mon.c.smithi012.stderr: 2: (std::_Rb_tree_rebalance_for_erase(std::_Rb_tree_node_base*, std::_Rb_tree_node_base&)+0x4) [0x7fcc7aec6514]
2024-03-08T01:00:45.506 INFO:tasks.ceph.mon.c.smithi012.stderr: 3: (CommonSafeTimer<std::mutex>::cancel_event(Context*)+0xb3) [0x7fcc7b8ad623]
2024-03-08T01:00:45.506 INFO:tasks.ceph.mon.c.smithi012.stderr: 4: (Monitor::cancel_probe_timeout()+0x51) [0x5565288a9661]
2024-03-08T01:00:45.506 INFO:tasks.ceph.mon.c.smithi012.stderr: 5: (Monitor::_reset()+0x70) [0x5565288a9870]
2024-03-08T01:00:45.506 INFO:tasks.ceph.mon.c.smithi012.stderr: 6: (Monitor::start_election()+0x55) [0x5565288aae25]
2024-03-08T01:00:45.506 INFO:tasks.ceph.mon.c.smithi012.stderr: 7: (Monitor::handle_probe_reply(boost::intrusive_ptr<MonOpRequest>)+0xe2b) [0x5565288ad57b]
2024-03-08T01:00:45.507 INFO:tasks.ceph.mon.c.smithi012.stderr: 8: (Monitor::handle_probe(boost::intrusive_ptr<MonOpRequest>)+0x211) [0x5565288ae7b1]
2024-03-08T01:00:45.507 INFO:tasks.ceph.mon.c.smithi012.stderr: 9: (Monitor::dispatch_op(boost::intrusive_ptr<MonOpRequest>)+0xe8a) [0x5565288d0dba]
2024-03-08T01:00:45.507 INFO:tasks.ceph.mon.c.smithi012.stderr: 10: (Monitor::_ms_dispatch(Message*)+0x7bf) [0x5565288c565f]
2024-03-08T01:00:45.507 INFO:tasks.ceph.mon.c.smithi012.stderr: 11: ceph-mon(+0x29950c) [0x55652888550c]
2024-03-08T01:00:45.507 INFO:tasks.ceph.mon.c.smithi012.stderr: 12: (DispatchQueue::entry()+0x52a) [0x7fcc7b9d70aa]
2024-03-08T01:00:45.507 INFO:tasks.ceph.mon.c.smithi012.stderr: 13: /usr/lib64/ceph/libceph-common.so.2(+0x467d01) [0x7fcc7ba67d01]
2024-03-08T01:00:45.507 INFO:tasks.ceph.mon.c.smithi012.stderr: 14: /lib64/libc.so.6(+0x9f802) [0x7fcc7aa9f802]
2024-03-08T01:00:45.507 INFO:tasks.ceph.mon.c.smithi012.stderr: 15: /lib64/libc.so.6(+0x3f450) [0x7fcc7aa3f450]
2024-03-08T01:00:45.507 INFO:tasks.ceph.mon.a.smithi012.stderr: ceph version 19.0.0-1910-g8d113723 (8d113723f96506c76b2d749e6808af82c9613dd5) squid (dev)
2024-03-08T01:00:45.508 INFO:tasks.ceph.mon.a.smithi012.stderr: 1: /lib64/libc.so.6(+0x54db0) [0x7ff09ac54db0]
2024-03-08T01:00:45.508 INFO:tasks.ceph.mon.a.smithi012.stderr: 2: (std::_Rb_tree_rebalance_for_erase(std::_Rb_tree_node_base*, std::_Rb_tree_node_base&)+0x4) [0x7ff09b0c6514]
2024-03-08T01:00:45.508 INFO:tasks.ceph.mon.a.smithi012.stderr: 3: (CommonSafeTimer<std::mutex>::cancel_event(Context*)+0xb3) [0x7ff09baad623]
2024-03-08T01:00:45.508 INFO:tasks.ceph.mon.a.smithi012.stderr: 4: (Monitor::cancel_probe_timeout()+0x51) [0x5584afa09661]
2024-03-08T01:00:45.508 INFO:tasks.ceph.mon.a.smithi012.stderr: 5: (Monitor::_reset()+0x70) [0x5584afa09870]
2024-03-08T01:00:45.508 INFO:tasks.ceph.mon.a.smithi012.stderr: 6: (Monitor::start_election()+0x55) [0x5584afa0ae25]
2024-03-08T01:00:45.508 INFO:tasks.ceph.mon.a.smithi012.stderr: 7: (Monitor::handle_probe_reply(boost::intrusive_ptr<MonOpRequest>)+0xe2b) [0x5584afa0d57b]
2024-03-08T01:00:45.508 INFO:tasks.ceph.mon.a.smithi012.stderr: 8: (Monitor::handle_probe(boost::intrusive_ptr<MonOpRequest>)+0x211) [0x5584afa0e7b1]
2024-03-08T01:00:45.508 INFO:tasks.ceph.mon.a.smithi012.stderr: 9: (Monitor::dispatch_op(boost::intrusive_ptr<MonOpRequest>)+0xe8a) [0x5584afa30dba]
2024-03-08T01:00:45.508 INFO:tasks.ceph.mon.a.smithi012.stderr: 10: (Monitor::_ms_dispatch(Message*)+0x7bf) [0x5584afa2565f]
2024-03-08T01:00:45.508 INFO:tasks.ceph.mon.a.smithi012.stderr: 11: ceph-mon(+0x29950c) [0x5584af9e550c]
2024-03-08T01:00:45.508 INFO:tasks.ceph.mon.a.smithi012.stderr: 12: (DispatchQueue::entry()+0x52a) [0x7ff09bbd70aa]
2024-03-08T01:00:45.508 INFO:tasks.ceph.mon.a.smithi012.stderr: 13: /usr/lib64/ceph/libceph-common.so.2(+0x467d01) [0x7ff09bc67d01]
2024-03-08T01:00:45.509 INFO:tasks.ceph.mon.a.smithi012.stderr: 14: /lib64/libc.so.6(+0x9f802) [0x7ff09ac9f802]
2024-03-08T01:00:45.509 INFO:tasks.ceph.mon.a.smithi012.stderr: 15: /lib64/libc.so.6(+0x3f450) [0x7ff09ac3f450]
2024-03-08T01:00:45.509 INFO:tasks.ceph.mon.c.smithi012.stderr:2024-03-08T01:00:45.505+0000 7fcc72ed6640 -1 *** Caught signal (Segmentation fault) **
2024-03-08T01:00:45.509 INFO:tasks.ceph.mon.c.smithi012.stderr: in thread 7fcc72ed6640 thread_name:ms_dispatch
```

1/4 looks like

```
2024-03-08T01:05:59.458 INFO:tasks.ceph.mgr.x.smithi122.stderr:*** Caught signal (Segmentation fault) **
2024-03-08T01:05:59.458 INFO:tasks.ceph.mgr.x.smithi122.stderr: in thread 7fcbcd22c040 thread_name:ceph-mgr
2024-03-08T01:05:59.459 INFO:tasks.ceph.mgr.x.smithi122.stderr: ceph version 19.0.0-1910-g8d113723 (8d113723f96506c76b2d749e6808af82c9613dd5) squid (dev)
2024-03-08T01:05:59.459 INFO:tasks.ceph.mgr.x.smithi122.stderr: 1: /lib64/libc.so.6(+0x54db0) [0x7fcbccc54db0]
2024-03-08T01:05:59.459 INFO:tasks.ceph.mgr.x.smithi122.stderr: 2: (std::_Rb_tree_rebalance_for_erase(std::_Rb_tree_node_base*, std::_Rb_tree_node_base&)+0x4) [0x7fcbcd0c6514]
2024-03-08T01:05:59.459 INFO:tasks.ceph.mgr.x.smithi122.stderr: 3: (CommonSafeTimer<std::mutex>::cancel_all_events()+0x85) [0x7fcbcd6ad8d5]
2024-03-08T01:05:59.459 INFO:tasks.ceph.mgr.x.smithi122.stderr: 4: (CommonSafeTimer<std::mutex>::shutdown()+0x3d) [0x7fcbcd6adb8d]
2024-03-08T01:05:59.459 INFO:tasks.ceph.mgr.x.smithi122.stderr: 5: (MonClient::shutdown()+0x3a9) [0x7fcbcd8eff29]
2024-03-08T01:05:59.459 INFO:tasks.ceph.mgr.x.smithi122.stderr: 6: (MonClient::get_monmap_and_config()+0x5e5) [0x7fcbcd8f1a85]
```

athanatos commented:
This...might be a gcc lto bug -- https://bugzilla.redhat.com/show_bug.cgi?id=2241339

I'm going to add a commit disabling lto and rerun it.

athanatos commented:
wip-sjust-testing-2024-03-08

athanatos commented:
I added the following patch to remove lto:

```
commit 42f37d9bfbbd6938bfdf54efc04c268e9b64e05c (HEAD -> sjust/wip-disable-lto, origin/sjust/wip-disable-lto)
Author: Samuel Just <sjust@redhat.com>
Date:   Fri Mar 8 04:03:03 2024 +0000

    DNM: ceph.spec.in: disable lto

    Fixes: https://tracker.ceph.com/issues/64804
    Signed-off-by: Samuel Just <sjust@redhat.com>

diff --git a/ceph.spec.in b/ceph.spec.in
index ed367465527..b1f9c125c57 100644
--- a/ceph.spec.in
+++ b/ceph.spec.in
@@ -1331,9 +1331,7 @@ This package provides a Ceph hardware monitoring agent.
 %build
 # Disable lto on systems that do not support symver attribute
 # See https://gcc.gnu.org/bugzilla/show_bug.cgi?id=48200 for details
-%if ( 0%{?rhel} && 0%{?rhel} < 9 ) || ( 0%{?suse_version} && 0%{?suse_version} <= 1500 )
 %define _lto_cflags %{nil}
-%endif

 %if 0%{with cephfs_java}
 # Find jni.h
@@ -1353,10 +1351,13 @@ export LDFLAGS="$RPM_LD_FLAGS"

 %if 0%{with seastar}
 # seastar uses longjmp() to implement coroutine. and this annoys longjmp_chk()
-export CXXFLAGS=$(echo $RPM_OPT_FLAGS | sed -e 's/-Wp,-D_FORTIFY_SOURCE=2//g')
+export CXXFLAGS=$(echo $RPM_OPT_FLAGS | sed -e 's/-Wp,-D_FORTIFY_SOURCE=2//g' | sed -e 's/-flto=auto -ffat-lto-objects//g')
 # remove from CFLAGS too because it causes the arrow submodule to fail with:
 #   warning _FORTIFY_SOURCE requires compiling with optimization (-O)
-export CFLAGS=$(echo $RPM_OPT_FLAGS | sed -e 's/-Wp,-D_FORTIFY_SOURCE=2//g')
+export CFLAGS=$(echo $RPM_OPT_FLAGS | sed -e 's/-Wp,-D_FORTIFY_SOURCE=2//g' | sed -e 's/-flto=auto -ffat-lto-objects//g')
+%else
+export CXXFLAGS=$(echo $RPM_OPT_FLAGS | sed -e 's/-flto=auto -ffat-lto-objects//g')
+export CFLAGS=$(echo $RPM_OPT_FLAGS | sed -e 's/-flto=auto -ffat-lto-objects//g')
 %endif

 env | sort
```

This seems to have caused a linker error for reasons I do not yet understand:

```
cd /home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos9/DIST/centos9/MACHINE_SIZE/gigantic/release/19.0.0-1977-gb0c1f7e8/rpm/el9/BUILD/ceph-19.0.0-1977-gb0c1f7e8/redhat-linux-build/src/tools && /usr/bin/cmake -E cmake_link_script CMakeFiles/ceph_scratchtool.dir/link.txt --verbose=1
/opt/rh/gcc-toolset-13/root/usr/bin/c++ -O2  -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -m64 -march=x86-64-v2 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -Wl,-z,relro -Wl,--as-needed  -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld -rdynamic -pie CMakeFiles/ceph_scratchtool.dir/scratchtool.c.o -o ../../bin/ceph_scratchtool  -Wl,-rpath,/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos9/DIST/centos9/MACHINE_SIZE/gigantic/release/19.0.0-1977-gb0c1f7e8/rpm/el9/BUILD/ceph-19.0.0-1977-gb0c1f7e8/redhat-linux-build/lib: ../../lib/librados.so.2.0.0 ../../lib/libglobal.a ../../lib/libceph-common.so.2 ../../lib/libjson_spirit.a ../../lib/libcommon_utf8.a ../../lib/liberasure_code.a ../../lib/libextblkdev.a -lcap ../../boost/lib/libboost_thread.a ../../boost/lib/libboost_chrono.a ../../boost/lib/libboost_atomic.a ../../boost/lib/libboost_system.a ../../boost/lib/libboost_random.a ../../boost/lib/libboost_program_options.a ../../boost/lib/libboost_date_time.a ../../boost/lib/libboost_iostreams.a ../../boost/lib/libboost_regex.a /usr/lib64/libblkid.so /usr/lib64/libcrypto.so /usr/lib64/libudev.so /usr/lib64/libibverbs.so /usr/lib64/librdmacm.so /usr/lib64/libz.so ../opentelemetry-cpp/sdk/src/trace/libopentelemetry_trace.a ../opentelemetry-cpp/sdk/src/resource/libopentelemetry_resources.a ../opentelemetry-cpp/sdk/src/common/libopentelemetry_common.a ../opentelemetry-cpp/exporters/jaeger/libopentelemetry_exporter_jaeger_trace.a ../opentelemetry-cpp/ext/src/http/client/curl/libopentelemetry_http_client_curl.a /usr/lib64/libcurl.so /usr/lib64/libthrift.so -ldl /usr/lib64/librt.a -lresolv  -Wl,--as-needed -latomic
/opt/rh/gcc-toolset-13/root/usr/libexec/gcc/x86_64-redhat-linux/13/ld: /opt/rh/gcc-toolset-13/root/usr/lib/gcc/x86_64-redhat-linux/13/libstdc++_nonshared.a(ios_init.o): undefined reference to symbol '_ZNSt8ios_base4InitD1Ev@@GLIBCXX_3.4'
/opt/rh/gcc-toolset-13/root/usr/libexec/gcc/x86_64-redhat-linux/13/ld: /usr/lib64/libstdc++.so.6: error adding symbols: DSO missing from command line
```

The corresponding line from a successfull build:

```
/opt/rh/gcc-toolset-13/root/usr/bin/c++ -O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong   -m64 -march=x86-64-v2 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -Wl,-z,relro -Wl,--as-needed  -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld -rdynamic -pie CMakeFiles/ceph_scratchtool.dir/scratchtool.c.o -o ../../bin/ceph_scratchtool  -Wl,-rpath,/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos9/DIST/centos9/MACHINE_SIZE/gigantic/release/19.0.0-1944-gd9866ed1/rpm/el9/BUILD/ceph-19.0.0-1944-gd9866ed1/redhat-linux-build/lib: ../../lib/librados.so.2.0.0 ../../lib/libglobal.a ../../lib/libceph-common.so.2 ../../lib/libjson_spirit.a ../../lib/libcommon_utf8.a ../../lib/liberasure_code.a ../../lib/libextblkdev.a -lcap ../../boost/lib/libboost_thread.a ../../boost/lib/libboost_chrono.a ../../boost/lib/libboost_atomic.a ../../boost/lib/libboost_system.a ../../boost/lib/libboost_random.a ../../boost/lib/libboost_program_options.a ../../boost/lib/libboost_date_time.a ../../boost/lib/libboost_iostreams.a ../../boost/lib/libboost_regex.a /usr/lib64/libblkid.so /usr/lib64/libcrypto.so /usr/lib64/libudev.so /usr/lib64/libibverbs.so /usr/lib64/librdmacm.so /usr/lib64/libz.so ../opentelemetry-cpp/sdk/src/trace/libopentelemetry_trace.a ../opentelemetry-cpp/sdk/src/resource/libopentelemetry_resources.a ../opentelemetry-cpp/sdk/src/common/libopentelemetry_common.a ../opentelemetry-cpp/exporters/jaeger/libopentelemetry_exporter_jaeger_trace.a ../opentelemetry-cpp/ext/src/http/client/curl/libopentelemetry_http_client_curl.a /usr/lib64/libcurl.so /usr/lib64/libthrift.so -ldl /usr/lib64/librt.a -lresolv  -Wl,--as-needed-latomic
```

Only difference seems to be the -flto=auto -ffat-lto-objects bit.

athanatos commented:
Putting ```%define _lto_cflags %{nil}``` at the top of the ceph.spec.in file actually does disable lto -- it just doesn't work if it's at the start of the %build section where current main has it conditionally for systems that don't support the symver attribute.

Still chasing down the problem I mention above, though.

tchaikov commented:
i am also trying to understand the root cause. added `-lstdc++` to `LDFLAGS`, but it does not help:
```
[ 76%] Linking CXX executable ../../../bin/ceph_test_trim_caps
cd /home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos9/DIST/centos9/MACHINE_SIZE/gigantic/release/19.0.0-1981-g7acdef4b/rpm/el9/BUILD/ceph-19.0.0-1981-g7acdef4b/redhat-linux-build/src/test/fs && /usr/bin/cmake -E cmake_link_script CMakeFiles/ceph_test_trim_caps.dir/link.txt --verbose=1
/opt/rh/gcc-toolset-13/root/usr/bin/c++ -O2  -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong   -m64 -march=x86-64-v2 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -Wl,-z,relro -Wl,--as-needed  -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld   -lstdc++ -rdynamic -pie CMakeFiles/ceph_test_trim_caps.dir/test_trim_caps.cc.o -o ../../../bin/ceph_test_trim_caps  -Wl,-rpath,/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos9/DIST/centos9/MACHINE_SIZE/gigantic/release/19.0.0-1981-g7acdef4b/rpm/el9/BUILD/ceph-19.0.0-1981-g7acdef4b/redhat-linux-build/lib: ../../../lib/libcephfs.so.2.0.0 ../../../lib/libceph-common.so.2 ../../../lib/libjson_spirit.a ../../../lib/libcommon_utf8.a ../../../lib/liberasure_code.a ../../../lib/libextblkdev.a -lcap -lresolv ../../../boost/lib/libboost_thread.a ../../../boost/lib/libboost_chrono.a ../../../boost/lib/libboost_atomic.a ../../../boost/lib/libboost_system.a ../../../boost/lib/libboost_random.a ../../../boost/lib/libboost_program_options.a ../../../boost/lib/libboost_date_time.a ../../../boost/lib/libboost_iostreams.a ../../../boost/lib/libboost_regex.a /usr/lib64/libblkid.so /usr/lib64/libcrypto.so -ldl /usr/lib64/libudev.so /usr/lib64/libibverbs.so /usr/lib64/librdmacm.so /usr/lib64/libz.so ../../opentelemetry-cpp/sdk/src/trace/libopentelemetry_trace.a ../../opentelemetry-cpp/sdk/src/resource/libopentelemetry_resources.a ../../opentelemetry-cpp/sdk/src/common/libopentelemetry_common.a ../../opentelemetry-cpp/exporters/jaeger/libopentelemetry_exporter_jaeger_trace.a ../../opentelemetry-cpp/ext/src/http/client/curl/libopentelemetry_http_client_curl.a /usr/lib64/libcurl.so /usr/lib64/libthrift.so  -Wl,--as-needed -latomic 
/opt/rh/gcc-toolset-13/root/usr/libexec/gcc/x86_64-redhat-linux/13/ld: /opt/rh/gcc-toolset-13/root/usr/lib/gcc/x86_64-redhat-linux/13/libstdc++_nonshared.a(ios_init.o): undefined reference to symbol '_ZNSt8ios_base4InitD1Ev@@GLIBCXX_3.4'
/opt/rh/gcc-toolset-13/root/usr/libexec/gcc/x86_64-redhat-linux/13/ld: /usr/lib64/libstdc++.so.6: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
make[2]: *** [src/test/fs/CMakeFiles/ceph_test_trim_caps.dir/build.make:125: bin/ceph_test_trim_caps] Error 1
```

see https://jenkins.ceph.com/job/ceph-dev-new-build/ARCH=x86_64,AVAILABLE_ARCH=x86_64,AVAILABLE_DIST=centos9,DIST=centos9,MACHINE_SIZE=gigantic/77463//consoleFull

athanatos commented:
https://jenkins.ceph.com/job/ceph-dev-new-build/ARCH=arm64,AVAILABLE_ARCH=arm64,AVAILABLE_DIST=centos9,DIST=centos9,MACHINE_SIZE=gigantic/77463/consoleFull#-213514296085e1414f-af06-4588-8fed-ab5891ddc58a

Removing ceph_scratchtool just causes it to fail on a later linker invocation in the same way:

```
/opt/rh/gcc-toolset-13/root/usr/bin/c++ -O2  -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong   -fasynchronous-unwind-tables -fstack-clash-protection -Wl,-z,relro -Wl,--as-needed  -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld -rdynamic -pie CMakeFiles/ceph_test_trim_caps.dir/test_trim_caps.cc.o -o ../../../bin/ceph_test_trim_caps  -Wl,-rpath,/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/arm64/AVAILABLE_ARCH/arm64/AVAILABLE_DIST/centos9/DIST/centos9/MACHINE_SIZE/gigantic/release/19.0.0-1980-g4eafb9de/rpm/el9/BUILD/ceph-19.0.0-1980-g4eafb9de/redhat-linux-build/lib: ../../../lib/libcephfs.so.2.0.0 ../../../lib/libceph-common.so.2 ../../../lib/libjson_spirit.a ../../../lib/libcommon_utf8.a ../../../lib/liberasure_code.a ../../../lib/libextblkdev.a -lcap -lresolv ../../../boost/lib/libboost_thread.a ../../../boost/lib/libboost_chrono.a ../../../boost/lib/libboost_atomic.a ../../../boost/lib/libboost_system.a ../../../boost/lib/libboost_random.a ../../../boost/lib/libboost_program_options.a ../../../boost/lib/libboost_date_time.a ../../../boost/lib/libboost_iostreams.a ../../../boost/lib/libboost_regex.a /usr/lib64/libblkid.so /usr/lib64/libcrypto.so -ldl /usr/lib64/libudev.so /usr/lib64/libibverbs.so /usr/lib64/librdmacm.so /usr/lib64/libz.so ../../opentelemetry-cpp/sdk/src/trace/libopentelemetry_trace.a ../../opentelemetry-cpp/sdk/src/resource/libopentelemetry_resources.a ../../opentelemetry-cpp/sdk/src/common/libopentelemetry_common.a ../../opentelemetry-cpp/exporters/jaeger/libopentelemetry_exporter_jaeger_trace.a ../../opentelemetry-cpp/ext/src/http/client/curl/libopentelemetry_http_client_curl.a /usr/lib64/libcurl.so /usr/lib64/libthrift.so  -Wl,--as-needed -latomic 
/opt/rh/gcc-toolset-13/root/usr/libexec/gcc/aarch64-redhat-linux/13/ld: /opt/rh/gcc-toolset-13/root/usr/lib/gcc/aarch64-redhat-linux/13/libstdc++_nonshared.a(ios_init.o): undefined reference to symbol '_ZNSt8ios_base4InitD1Ev@@GLIBCXX_3.4'
/opt/rh/gcc-toolset-13/root/usr/libexec/gcc/aarch64-redhat-linux/13/ld: /usr/lib64/libstdc++.so.6: error adding symbols: DSO missing from command line
```

tchaikov commented:
in the test, kernel page cache is used, so ops are performed out of the reactor thread, instead, `io_submit()` is called in the per-reactor thread pool. but in the thread pool, we only read from `_aio_retries`, and never mutates it. in other words, the only thread mutates `_aio_retries` is the reactor thread.

```mermaid
stateDiagram-v2
    iocb_pool --> submission_queue: if iocb_pool is not empty
    submission_queue --> pending_aio_retry: if opts_kernel_page_cache
    pending_aio_retry --> aio_retries: call io_submit() in thread pool
    aio_retries --> iocb_pool: io_submit() returns EBADF
   aio_retries --> aio_event: 
   aio_event --> pending_aio_retry: EAGAIN
   aio_event --> iocb_pool: handled
```

athanatos commented:
@tchaikov I think ^ was probalby meant for another PR?

tchaikov commented:
sorry, was looking at a seastar issue. removed.

athanatos commented:
I got it to reproduce locally.

```
~/rpmbuild » /ceph-19.0.0-1980-gd5586a6d16c/redhat-linux-build/src/include -I/home/sam/rpmbuild/BUILD/ceph-19.0.0-1980-gd5586a6d16c/src -isystem /home/sam/rpmbuild/BUILD/ceph-19.0.0-1980-gd5586a6d16c/redhat-linux-build/boost/include -isystem /home/sam/rpmbuild/BUILD/ceph-19.0.0-1980-gd5586a6d16c/redhat-linux-build/include -isystem /home/sam/rpmbuild/BUILD/ceph-19.0.0-1980-gd5586a6d16c/src/jaegertracing/opentelemetry-cpp/api/include -isystem /home/sam/rpmbuild/BUILD/ceph-19.0.0-1980-gd5586a6d16c/src/jaegertracing/opentelemetry-cpp/exporters/jaeger/include -isystem /home/sam/rpmbuild/BUILD/ceph-19.0.0-1980-gd5586a6d16c/src/jaegertracing/opentelemetry-cpp/ext/include -isystem /home/sam/rpmbuild/BUILD/ceph-19.0.0-1980-gd5586a6d16c/src/jaegertracing/opentelemetry-cpp/sdk/include -isystem /home/sam/rpmbuild/BUILD/ceph-19.0.0-1980-gd5586a6d16c/src/xxHash -isystem /home/sam/rpmbuild/BUILD/ceph-19.0.0-1980-gd5586a6d16c/src/fmt/include -O2  -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong   -m64 -march=x86-64-v2 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -std=c++20 -fPIE   -U_FORTIFY_SOURCE -fno-builtin-malloc -fno-builtin-calloc -fno-builtin-realloc -fno-builtin-free -DBOOST_PHOENIX_STL_TUPLE_H_ -Wall -fno-strict-aliasing -fsigned-char -Wtype-limits -Wignored-qualifiers -Wpointer-arith -Werror=format-security -Winit-self -Wno-unknown-pragmas -Wnon-virtual-dtor -Wno-ignored-qualifiers -ftemplate-depth-1024 -Wpessimizing-move -Wredundant-move -Wstrict-null-sentinel -Woverloaded-virtual -fstack-protector-strong -fdiagnostics-color=auto -MD -MT src/test/fs/CMakeFiles/ceph_test_trim_caps.dir/test_trim_caps.cc.o -MF CMakeFiles/ceph_test_trim_caps.dir/test_trim_caps.cc.o.d -o CMakeFiles/ceph_test_trim_caps.dir/test_trim_caps.cc.o -c /home/sam/rpmbuild/BUILD/ceph-19.0.0-1980-gd5586a6d16c/src/test/fs/test_trim_caps.cc
~/rpmbuild » cd /home/sam/rpmbuild/BUILD/ceph-19.0.0-1980-gd5586a6d16c/redhat-linux-build/src/test/fs && /opt/rh/gcc-toolset-13/root/usr/bin/c++ -DBOOST_ASIO_DISABLE_THREAD_KEYWORD_EXTENSION -DBOOST_ASIO_HAS_IO_URING -DBOOST_ASIO_NO_TS_EXECUTORS -DFMT_HEADER_ONLY=1 -DHAVE_CONFIG_H -D_FILE_OFFSET_BITS=64 -D_FORTIFY_SOURCE=2 -D_GNU_SOURCE -D_REENTRANT -D_THREAD_SAFE -D__CEPH__ -D__STDC_FORMAT_MACROS -D__linux__ -I/home/sam/rpmbuild/BUILD/ceph-19.0.0-1980-gd5586a6d16c/redhat-linux-build/src/include -I/home/sam/rpmbuild/BUILD/ceph-19.0.0-1980-gd5586a6d16c/src -isystem /home/sam/rpmbuild/BUILD/ceph-19.0.0-1980-gd5586a6d16c/redhat-linux-build/boost/include -isystem /home/sam/rpmbuild/BUILD/ceph-19.0.0-1980-gd5586a6d16c/redhat-linux-build/include -isystem /home/sam/rpmbuild/BUILD/ceph-19.0.0-1980-gd5586a6d16c/src/jaegertracing/opentelemetry-cpp/api/include -isystem /home/sam/rpmbuild/BUILD/ceph-19.0.0-1980-gd5586a6d16c/src/jaegertracing/opentelemetry-cpp/exporters/jaeger/include -isystem /home/sam/rpmbuild/BUILD/ceph-19.0.0-1980-gd5586a6d16c/src/jaegertracing/opentelemetry-cpp/ext/include -isystem /home/sam/rpmbuild/BUILD/ceph-19.0.0-1980-gd5586a6d16c/src/jaegertracing/opentelemetry-cpp/sdk/include -isystem /home/sam/rpmbuild/BUILD/ceph-19.0.0-1980-gd5586a6d16c/src/xxHash -isystem /home/sam/rpmbuild/BUILD/ceph-19.0.0-1980-gd5586a6d16c/src/fmt/include -O2  -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong   -m64 -march=x86-64-v2 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -std=c++20 -fPIE   -U_FORTIFY_SOURCE -fno-builtin-malloc -fno-builtin-calloc -fno-builtin-realloc -fno-builtin-free -DBOOST_PHOENIX_STL_TUPLE_H_ -Wall -fno-strict-aliasing -fsigned-char -Wtype-limits -Wignored-qualifiers -Wpointer-arith -Werror=format-security -Winit-self -Wno-unknown-pragmas -Wnon-virtual-dtor -Wno-ignored-qualifiers -ftemplate-depth-1024 -Wpessimizing-move -Wredundant-move -Wstrict-null-sentinel -Woverloaded-virtual -fstack-protector-strong -fdiagnostics-color=auto -MD -MT src/test/fs/CMakeFiles/ceph_test_trim_caps.dir/test_trim_caps.cc.o -MF CMakeFiles/ceph_test_trim_caps.dir/test_trim_caps.cc.o.d -o CMakeFiles/ceph_test_trim_caps.dir/test_trim_caps.cc.o -c /home/sam/rpmbuild/BUILD/ceph-19.0.0-1980-gd5586a6d16c/src/test/fs/test_trim_caps.cc
test/fs » ls
test/fs » cd /home/sam/rpmbuild/BUILD/ceph-19.0.0-1980-gd5586a6d16c/redhat-linux-build/src/test/fs && /opt/rh/gcc-toolset-13/root/usr/bin/c++ -DBOOST_ASIO_DISABLE_THREAD_KEYWORD_EXTENSION -DBOOST_ASIO_HAS_IO_URING -DBOOST_ASIO_NO_TS_EXECUTORS -DFMT_HEADER_ONLY=1 -DHAVE_CONFIG_H -D_FILE_OFFSET_BITS=64 -D_FORTIFY_SOURCE=2 -D_GNU_SOURCE -D_REENTRANT -D_THREAD_SAFE -D__CEPH__ -D__STDC_FORMAT_MACROS -D__linux__ -I/home/sam/rpmbuild/BUILD/ceph-19.0.0-1980-gd5586a6d16c/redhat-linux-build/src/include -I/home/sam/rpmbuild/BUILD/ceph-19.0.0-1980-gd5586a6d16c/src -isystem /home/sam/rpmbuild/BUILD/ceph-19.0.0-1980-gd5586a6d16c/redhat-linux-build/boost/include -isystem /home/sam/rpmbuild/BUILD/ceph-19.0.0-1980-gd5586a6d16c/redhat-linux-build/include -isystem /home/sam/rpmbuild/BUILD/ceph-19.0.0-1980-gd5586a6d16c/src/jaegertracing/opentelemetry-cpp/api/include -isystem /home/sam/rpmbuild/BUILD/ceph-19.0.0-1980-gd5586a6d16c/src/jaegertracing/opentelemetry-cpp/exporters/jaeger/include -isystem /home/sam/rpmbuild/BUILD/ceph-19.0.0-1980-gd5586a6d16c/src/jaegertracing/opentelemetry-cpp/ext/include -isystem /home/sam/rpmbuild/BUILD/ceph-19.0.0-1980-gd5586a6d16c/src/jaegertracing/opentelemetry-cpp/sdk/include -isystem /home/sam/rpmbuild/BUILD/ceph-19.0.0-1980-gd5586a6d16c/src/xxHash -isystem /home/sam/rpmbuild/BUILD/ceph-19.0.0-1980-gd5586a6d16c/src/fmt/include -O2  -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong   -m64 -march=x86-64-v2 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -std=c++20 -fPIE   -U_FORTIFY_SOURCE -fno-builtin-malloc -fno-builtin-calloc -fno-builtin-realloc -fno-builtin-free -DBOOST_PHOENIX_STL_TUPLE_H_ -Wall -fno-strict-aliasing -fsigned-char -Wtype-limits -Wignored-qualifiers -Wpointer-arith -Werror=format-security -Winit-self -Wno-unknown-pragmas -Wnon-virtual-dtor -Wno-ignored-qualifiers -ftemplate-depth-1024 -Wpessimizing-move -Wredundant-move -Wstrict-null-sentinel -Woverloaded-virtual -fstack-protector-strong -fdiagnostics-color=auto -MD -MT src/test/fs/CMakeFiles/ceph_test_trim_caps.dir/test_trim_caps.cc.o -MF CMakeFiles/ceph_test_trim_caps.dir/test_trim_caps.cc.o.d -o CMakeFiles/ceph_test_trim_caps.dir/test_trim_caps.cc.o -c /home/sam/rpmbuild/BUILD/ceph-19.0.0-1980-gd5586a6d16c/src/test/fs/test_trim_caps.cc
test/fs » rm
test/fs » cd /home/sam/rpmbuild/BUILD/ceph-19.0.0-1980-gd5586a6d16c/redhat-linux-build/src/test/fs
test/fs » rm CMakeFiles/ceph_test_trim_caps.dir/test_trim_caps.cc.o
test/fs » cd /home/sam/rpmbuild/BUILD/ceph-19.0.0-1980-gd5586a6d16c/redhat-linux-build/src/test/fs && /opt/rh/gcc-toolset-13/root/usr/bin/c++ -DBOOST_ASIO_DISABLE_THREAD_KEYWORD_EXTENSION -DBOOST_ASIO_HAS_IO_URING -DBOOST_ASIO_NO_TS_EXECUTORS -DFMT_HEADER_ONLY=1 -DHAVE_CONFIG_H -D_FILE_OFFSET_BITS=64 -D_FORTIFY_SOURCE=2 -D_GNU_SOURCE -D_REENTRANT -D_THREAD_SAFE -D__CEPH__ -D__STDC_FORMAT_MACROS -D__linux__ -I/home/sam/rpmbuild/BUILD/ceph-19.0.0-1980-gd5586a6d16c/redhat-linux-build/src/include -I/home/sam/rpmbuild/BUILD/ceph-19.0.0-1980-gd5586a6d16c/src -isystem /home/sam/rpmbuild/BUILD/ceph-19.0.0-1980-gd5586a6d16c/redhat-linux-build/boost/include -isystem /home/sam/rpmbuild/BUILD/ceph-19.0.0-1980-gd5586a6d16c/redhat-linux-build/include -isystem /home/sam/rpmbuild/BUILD/ceph-19.0.0-1980-gd5586a6d16c/src/jaegertracing/opentelemetry-cpp/api/include -isystem /home/sam/rpmbuild/BUILD/ceph-19.0.0-1980-gd5586a6d16c/src/jaegertracing/opentelemetry-cpp/exporters/jaeger/include -isystem /home/sam/rpmbuild/BUILD/ceph-19.0.0-1980-gd5586a6d16c/src/jaegertracing/opentelemetry-cpp/ext/include -isystem /home/sam/rpmbuild/BUILD/ceph-19.0.0-1980-gd5586a6d16c/src/jaegertracing/opentelemetry-cpp/sdk/include -isystem /home/sam/rpmbuild/BUILD/ceph-19.0.0-1980-gd5586a6d16c/src/xxHash -isystem /home/sam/rpmbuild/BUILD/ceph-19.0.0-1980-gd5586a6d16c/src/fmt/include -O2  -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong   -m64 -march=x86-64-v2 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -std=c++20 -fPIE   -U_FORTIFY_SOURCE -fno-builtin-malloc -fno-builtin-calloc -fno-builtin-realloc -fno-builtin-free -DBOOST_PHOENIX_STL_TUPLE_H_ -Wall -fno-strict-aliasing -fsigned-char -Wtype-limits -Wignored-qualifiers -Wpointer-arith -Werror=format-security -Winit-self -Wno-unknown-pragmas -Wnon-virtual-dtor -Wno-ignored-qualifiers -ftemplate-depth-1024 -Wpessimizing-move -Wredundant-move -Wstrict-null-sentinel -Woverloaded-virtual -fstack-protector-strong -fdiagnostics-color=auto -MD -MT src/test/fs/CMakeFiles/ceph_test_trim_caps.dir/test_trim_caps.cc.o -MF CMakeFiles/ceph_test_trim_caps.dir/test_trim_caps.cc.o.d -o CMakeFiles/ceph_test_trim_caps.dir/test_trim_caps.cc.o -c /home/sam/rpmbuild/BUILD/ceph-19.0.0-1980-gd5586a6d16c/src/test/fs/test_trim_caps.cc
test/fs » ls CMakeFiles/ceph_test_trim_caps.dir/test_trim_caps.cc.o
CMakeFiles/ceph_test_trim_caps.dir/test_trim_caps.cc.o
test/fs » /opt/rh/gcc-toolset-13/root/usr/bin/c++ -O2  -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong   -m64 -march=x86-64-v2 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -Wl,-z,relro -Wl,--as-needed  -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld -rdynamic -pie CMakeFiles/ceph_test_trim_caps.dir/test_trim_caps.cc.o -o ../../../bin/ceph_test_trim_caps  -Wl,-rpath,/home/sam/rpmbuild/BUILD/ceph-19.0.0-1980-gd5586a6d16c/redhat-linux-build/lib: ../../../lib/libcephfs.so.2.0.0 ../../../lib/libceph-common.so.2 ../../../lib/libjson_spirit.a ../../../lib/libcommon_utf8.a ../../../lib/liberasure_code.a ../../../lib/libextblkdev.a -lcap -lresolv ../../../boost/lib/libboost_thread.a ../../../boost/lib/libboost_chrono.a ../../../boost/lib/libboost_atomic.a ../../../boost/lib/libboost_system.a ../../../boost/lib/libboost_random.a ../../../boost/lib/libboost_program_options.a ../../../boost/lib/libboost_date_time.a ../../../boost/lib/libboost_iostreams.a ../../../boost/lib/libboost_regex.a /usr/lib64/libblkid.so /usr/lib64/libcrypto.so -ldl /usr/lib64/libudev.so /usr/lib64/libibverbs.so /usr/lib64/librdmacm.so /usr/lib64/libz.so ../../opentelemetry-cpp/sdk/src/trace/libopentelemetry_trace.a ../../opentelemetry-cpp/sdk/src/resource/libopentelemetry_resources.a ../../opentelemetry-cpp/sdk/src/common/libopentelemetry_common.a ../../opentelemetry-cpp/exporters/jaeger/libopentelemetry_exporter_jaeger_trace.a ../../opentelemetry-cpp/ext/src/http/client/curl/libopentelemetry_http_client_curl.a /usr/lib64/libcurl.so /usr/lib64/libthrift.so  -Wl,--as-needed -latomic
/opt/rh/gcc-toolset-13/root/usr/libexec/gcc/x86_64-redhat-linux/13/ld: /opt/rh/gcc-toolset-13/root/usr/lib/gcc/x86_64-redhat-linux/13/libstdc++_nonshared.a(ios_init.o): undefined reference to symbol '_ZNSt8ios_base4InitD1Ev@@GLIBCXX_3.4'
/opt/rh/gcc-toolset-13/root/usr/libexec/gcc/x86_64-redhat-linux/13/ld: /usr/lib64/libstdc++.so.6: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
```

Adding -lstdc++ to the linker command actually did work:

```
test/fs » /opt/rh/gcc-toolset-13/root/usr/bin/c++ -O2  -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong   -m64 -march=x86-64-v2 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -Wl,-z,relro -Wl,--as-needed  -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld -rdynamic -pie CMakeFiles/ceph_test_trim_caps.dir/test_trim_caps.cc.o -o ../../../bin/ceph_test_trim_caps  -Wl,-rpath,/home/sam/rpmbuild/BUILD/ceph-19.0.0-1980-gd5586a6d16c/redhat-linux-build/lib: ../../../lib/libcephfs.so.2.0.0 ../../../lib/libceph-common.so.2 ../../../lib/libjson_spirit.a ../../../lib/libcommon_utf8.a ../../../lib/liberasure_code.a ../../../lib/libextblkdev.a -lcap -lresolv ../../../boost/lib/libboost_thread.a ../../../boost/lib/libboost_chrono.a ../../../boost/lib/libboost_atomic.a ../../../boost/lib/libboost_system.a ../../../boost/lib/libboost_random.a ../../../boost/lib/libboost_program_options.a ../../../boost/lib/libboost_date_time.a ../../../boost/lib/libboost_iostreams.a ../../../boost/lib/libboost_regex.a /usr/lib64/libblkid.so /usr/lib64/libcrypto.so -ldl /usr/lib64/libudev.so /usr/lib64/libibverbs.so /usr/lib64/librdmacm.so /usr/lib64/libz.so ../../opentelemetry-cpp/sdk/src/trace/libopentelemetry_trace.a ../../opentelemetry-cpp/sdk/src/resource/libopentelemetry_resources.a ../../opentelemetry-cpp/sdk/src/common/libopentelemetry_common.a ../../opentelemetry-cpp/exporters/jaeger/libopentelemetry_exporter_jaeger_trace.a ../../opentelemetry-cpp/ext/src/http/client/curl/libopentelemetry_http_client_curl.a /usr/lib64/libcurl.so /usr/lib64/libthrift.so  -Wl,--as-needed -latomic -lstdc++
test/fs »
```

tchaikov commented:
so, as always, order matters. seems we just need to put `-lstdc++` at *the end* of command line?

tchaikov commented:
i am passing `-DCMAKE_EXE_LINKER_FLAGS=-lstdc++` to cmake. pushed to ceph-ci as wip-kefu-testing-2024-03-10 .

athanatos commented:
```
/opt/rh/gcc-toolset-13/root/usr/bin/c++ -O2  -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong   -m64 -march=x86-64-v2 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -Wl,-z,relro -Wl,--as-needed  -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld -rdynamic -pie CMakeFiles/ceph_test_trim_caps.dir/test_trim_caps.cc.o -o ../../../bin/ceph_test_trim_caps -Wl,-rpath,/home/sam/rpmbuild/BUILD/ceph-19.0.0-1980-gd5586a6d16c/redhat-linux-build/lib: -lstdc++ ../../../lib/libcephfs.so.2.0.0 ../../../lib/libceph-common.so.2 ../../../lib/libjson_spirit.a ../../../lib/libcommon_utf8.a ../../../lib/liberasure_code.a ../../../lib/libextblkdev.a -lcap -lresolv ../../../boost/lib/libboost_thread.a ../../../boost/lib/libboost_chrono.a ../../../boost/lib/libboost_atomic.a ../../../boost/lib/libboost_system.a ../../../boost/lib/libboost_random.a ../../../boost/lib/libboost_program_options.a ../../../boost/lib/libboost_date_time.a ../../../boost/lib/libboost_iostreams.a ../../../boost/lib/libboost_regex.a /usr/lib64/libblkid.so /usr/lib64/libcrypto.so -ldl /usr/lib64/libudev.so /usr/lib64/libibverbs.so /usr/lib64/librdmacm.so /usr/lib64/libz.so ../../opentelemetry-cpp/sdk/src/trace/libopentelemetry_trace.a ../../opentelemetry-cpp/sdk/src/resource/libopentelemetry_resources.a ../../opentelemetry-cpp/sdk/src/common/libopentelemetry_common.a ../../opentelemetry-cpp/exporters/jaeger/libopentelemetry_exporter_jaeger_trace.a ../../opentelemetry-cpp/ext/src/http/client/curl/libopentelemetry_http_client_curl.a /usr/lib64/libcurl.so /usr/lib64/libthrift.so  -Wl,--as-needed -latomic
```

fails, but

```
/opt/rh/gcc-toolset-13/root/usr/bin/c++ -O2  -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong   -m64 -march=x86-64-v2 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -Wl,-z,relro -Wl,--as-needed  -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld -rdynamic -pie CMakeFiles/ceph_test_trim_caps.dir/test_trim_caps.cc.o -o ../../../bin/ceph_test_trim_caps -Wl,-rpath,/home/sam/rpmbuild/BUILD/ceph-19.0.0-1980-gd5586a6d16c/redhat-linux-build/lib: ../../../lib/libcephfs.so.2.0.0 -lstdc++ ../../../lib/libceph-common.so.2 ../../../lib/libjson_spirit.a ../../../lib/libcommon_utf8.a ../../../lib/liberasure_code.a ../../../lib/libextblkdev.a -lcap -lresolv ../../../boost/lib/libboost_thread.a ../../../boost/lib/libboost_chrono.a ../../../boost/lib/libboost_atomic.a ../../../boost/lib/libboost_system.a ../../../boost/lib/libboost_random.a ../../../boost/lib/libboost_program_options.a ../../../boost/lib/libboost_date_time.a ../../../boost/lib/libboost_iostreams.a ../../../boost/lib/libboost_regex.a /usr/lib64/libblkid.so /usr/lib64/libcrypto.so -ldl /usr/lib64/libudev.so /usr/lib64/libibverbs.so /usr/lib64/librdmacm.so /usr/lib64/libz.so ../../opentelemetry-cpp/sdk/src/trace/libopentelemetry_trace.a ../../opentelemetry-cpp/sdk/src/resource/libopentelemetry_resources.a ../../opentelemetry-cpp/sdk/src/common/libopentelemetry_common.a ../../opentelemetry-cpp/exporters/jaeger/libopentelemetry_exporter_jaeger_trace.a ../../opentelemetry-cpp/ext/src/http/client/curl/libopentelemetry_http_client_curl.a /usr/lib64/libcurl.so /usr/lib64/libthrift.so  -Wl,--as-needed -latomic
```

works.  It seems that -lstdc++ must be after ../../../lib/libcephfs.so.2.0.0.

tchaikov commented:
indeed. as linker always looks for the missing symbol from "foo" in the objects / libraries specified after "foo" in the command line.

changelog:

- add a commit which passes `-DCMAKE_EXE_LINKER_FLAGS=-lstdc++` if `rhel>=8`.

tchaikov commented:
being compiled at https://shaman.ceph.com/builds/ceph/wip-kefu-testing-2024-03-10/55b7b41d6801fa2fef1f6a5208f16655c3603b02/

tchaikov commented:
the windows build failed like
```
ninja: error: unknown target 'ceph_radosaclceph_test_admin_socket_output'
```

i don't think the change in the .spec.in file could impact the windows build.

tchaikov commented:
being tested at https://pulpito.ceph.com/kchai-2024-03-10_11:05:17-rados-wip-kefu-testing-2024-03-10-distro-default-smithi/

tchaikov commented:
jenkins test make check

tchaikov commented:
jenkins test api

athanatos commented:
I scheduled the above branch for 10 runs -- the previous issues didn't necessarily manifest in the first one.  I also updated https://github.com/athanatos/ceph/tree/sjust/wip-disable-lto to remove the commit removing ceph-scratchtool and updated the commit removing lto with an explanation and pointers to the relevant bugs.

The larger question is: is it worth disabling lto on classic builds, or should we only do this and the switch to gcc-13 for crimson builds?

tchaikov commented:
> I scheduled the above branch for 10 runs -- the linker issues didn't necessarily manifest in the first one. I also updated https://github.com/athanatos/ceph/tree/sjust/wip-disable-lto to remove the commit removing ceph-scratchtool and updated the commit removing lto with an explanation and pointers to the relevant bugs -- that should probably be included in this PR.
> 

i will cherry-pick this change into changset.

> http://pulpito.ceph.com/sjust-2024-03-11_00:25:03-rados-wip-kefu-testing-2024-03-10-distro-default-smithi/

great! thanks. all of these ten runs passed.

> 
> The larger question is: is it worth disabling lto on classic builds, or should we only do this and the switch to gcc-13 for crimson builds?

i don't know the answer without performing benchmarks on the classic builds with and without LTO. probably the safer approach is to switch to gcc-13 only for crimson builds. and use gts-13 after we have numbers supporting the switch?

tchaikov commented:
pushed to https://shaman.ceph.com/builds/ceph/wip-kefu-testing-2024-03-11/ba7cb0b09d7fcb3cb494e360da4d07fcf0f055bb/

tchaikov commented:
v2:

- pass `-DCMAKE_EXE_LINKER_FLAGS=-lstdc++ to cmake`.

v3:

- cherry-pick and adapt the change to disable lto when building crimson
- only use gcc-13 when building crimson add avoid the potential performance regression due to disabling lto.

tchaikov commented:
built at https://shaman.ceph.com/builds/ceph/wip-kefu-testing-2024-03-11/913d127c8116ef6f0066d75cfdbba3811939936d/

tchaikov commented:
being tested at https://pulpito.ceph.com/kchai-2024-03-11_15:32:56-rados-wip-kefu-testing-2024-03-11-distro-default-smithi/

tchaikov commented:
jenkins test make check

tchaikov commented:
@athanatos hi Sam, could you take another look? i think it's good to merge now. hopefully, all three tests will pass when you check them.

athanatos commented:
make check is failing, but for a reason present in other PRs that we're investigating.  I'll schedule this one for a crimson run and another 10 run classic run (one of the three failed, though for a reason I've seen elsewhere).

athanatos commented:
http://pulpito.ceph.com/sjust-2024-03-12_01:14:20-rados-wip-kefu-testing-2024-03-11-distro-default-smithi/
http://pulpito.ceph.com/sjust-2024-03-12_01:15:24-crimson-rados-wip-kefu-testing-2024-03-11-distro-default-smithi/

athanatos commented:
I think if the above seem ok, it's probably good to merge.  @rzarzynski @cbodley final look?

tchaikov commented:
jenkins test make check

athanatos commented:
https://pulpito.ceph.com/sjust-2024-03-12_01:15:24-crimson-rados-wip-kefu-testing-2024-03-11-distro-default-smithi/ mostly failed.  I'm traveling this week, so it may be a bit before I get a chance to look at the logs.

Svelar commented:
> wondering if we can parameterize the ceph-osd systemd service unit, so that it can launch crimson or the classic osd based on user's preference. but we need to persist a mark so user cannot switch to another osd after the the store is initialized. as the disk layouts of crimson and bluestore are different.

Could it be picked up in crimson weekly?

athanatos commented:
Building wip-sjust-cb2b48d7-2024-03-12 to confirm that the above failures are due to this patch, haven't had a chance to look into the logs yet.

rzarzynski commented:
Putting `DNM` due to the comment above.

athanatos commented:
I think the failures are unrelated to this PR -- main isn't particularly clean at the moment.  I'd like to clean up main before merging this though.

athanatos commented:
https://tracker.ceph.com/issues/64996 may be triggered (caused?) by this switch to gcc-13 -- investigating.

tchaikov commented:
> > wondering if we can parameterize the ceph-osd systemd service unit, so that it can launch crimson or the classic osd based on user's preference. but we need to persist a mark so user cannot switch to another osd after the the store is initialized. as the disk layouts of crimson and bluestore are different.
> 
> Could it be picked up in crimson weekly?

i think it might be answered by the attendees =)

tchaikov commented:
> I think the failures are unrelated to this PR -- main isn't particularly clean at the moment. I'd like to clean up main before merging this though.

@tchaikov hi Sam, have you started working on this? otherwise i'd like to take a look at `main.cc` this weekend. but i might be slow.

athanatos commented:
@tchaikov By main, I meant the crimson-rados suite on ceph/main, not main.cc.  I have a set of fixes in a test branch (wip-sjust-crimson-testing-2024-03-20) along with this PR.  It finally got through ci, so I just scheduled another suite run.  http://pulpito.ceph.com/sjust-2024-03-21_02:35:36-crimson-rados-wip-sjust-crimson-testing-2024-03-20-distro-default-smithi/  Once we get a clean-ish suite run with this PR, we can merge.

tchaikov commented:
@athanatos ahh, thanks. i misread your comment.

athanatos commented:
https://pulpito.ceph.com/sjust-2024-03-21_02:35:36-crimson-rados-wip-sjust-crimson-testing-2024-03-20-distro-default-smithi/

Failures unrelated, caused by existing known bugs.
